internal/pkg/api/handleCheckin.go

```diff
@@ -555,6 +561,56 @@ func parseMeta(zlog zerolog.Logger, agent *model.Agent, req *CheckinRequest) ([]
 	return outMeta, nil
 }
 
+func parseComponents(zlog zerolog.Logger, agent *model.Agent, req *CheckinRequest) ([]byte, error) {
+
+	// Quick comparison first; compare the JSON payloads.
+	// If the data is not consistently normalized, this short-circuit will not work.
+	if bytes.Equal(req.Components, agent.Components) {
+		zlog.Trace().Msg("quick comparing agent components data is equal")
+		return nil, nil
+	}
+
+	// Deserialize the request components data
+	var reqComponents interface{}
+	if len(req.Components) > 0 {
+		if err := json.Unmarshal(req.Components, &reqComponents); err != nil {
+			return nil, errors.Wrap(err, "parseComponents request")
+		}
+	}
+
+	// If empty, don't step on existing data
+	if reqComponents == nil {
+		return nil, nil
+	}
```

Comment:
@aleksmaus We should probably assert the data we are receiving from Elastic Agent before sending that to Elasticsearch? Maybe we should also put some kind of limit, maybe a size based limit on the `payload` field to reduce the risk or document size explosion?

Reply:
This approach is basically copied from parseMeta that is used currently for "local_metadata" handling which is another blob that agent can send whatever to the fleet server. 
There is already a limit check on the checking body size, default is ```defaultCheckinMaxBody  = 1024 * 1024```.
There is already the parsing of the components happens in the ```parseComponents``` which validates that is a parsable JSON. I can add a check if the components is an array.

Do we need to add field by field check? What are guarantees which fields are required vs which fields are optional?

Reply:
Will add fully structured parsing to the "components" property tomorrow, only things that are specified in the V2 at the moment will be stored for the components, and it will need to be extended for anything extra in the future.

Reply:
Yes, once everything is V2 we should be able to be more strict on any incoming data, not sure during the transition period we can do this.

Reply:
Sounds good. Will leave it as is for now until we stabilize the schema, can tighten validation later.